src/axom/quest/ShapeMesh.hpp

```diff
@@ -0,0 +1,343 @@
+   *   If empty, use the first material set in @c bpMesh.
+   *
+   * Incoming mesh array data are assumed to be accessible by the
+   * runtime policy, or an error will result.  (However the data need
+   * not correspond to the allocator id.)
+  */
+  ShapeMesh(RuntimePolicy runtimePolicy,
+            int allocatorId,
+            sidre::Group* bpMesh,
+            const std::string& topoName = {},
+            const std::string& matsetName = {});
+#endif
+
+  /*!
+   * @brief Runtime policy set in constructor.
+   */
+  RuntimePolicy getRuntimePolicy() const { return m_runtimePolicy; }
+
+  /*!
+   * @brief Allocator id set in constructor.
+   */
+  int getAllocatorID() const { return m_allocId; }
+
+  /*
+   * !@brief Return computational mesh as a sidre::Group if it has
+   * that form, or nullptr otherwise.
+  */
+  sidre::Group* getMeshAsSidre() { return m_bpGrpExt; }
+
+  /*!
+   * @brief Return computational mesh as a conduit::Node if it has
+   * that form, or nullptr otherwise.
+  */
+  conduit::Node* getMeshAsConduit() { return m_bpNodeExt; }
+
+  //!@brief Dimension of the mesh (2 or 3)
+  int dimension() const { return m_dim; }
+
+  //!@brief Number of cells in mesh.
+  IndexType getCellCount() const { return m_cellCount; }
+
+  //!@brief Number of vertices in mesh.
+  IndexType getVertexCount() const { return m_vertexCount; }
+
+  //!@brief Set the threshold to snapping vertex coordinates near
+  // zero to zero.  Default threshold is 1e-10.
+  void setZeroThreshold(double threshold)
+  {
+    m_zeroThreshold = threshold;
+  }
+
+  //@{
+  //!@name Accessors to mesh data.
+  //@}
+
+  //@{
+  //!@name Accessors to mesh-dependent intermediate data.
+  /*
+   * This data is dynamically generated as needed, and cached for use
+   * by multiple geometry clippers.  The idea is to eliminate redundant
+   * code and computations.
+   */
+  /*!
+   * @brief Tetrahedral version of mesh cells with cell \c i having tet ids in
+   * [i*NUM_TETS_PER_HEX, (i+1)*NUM_TETS_PER_HEX).
+   */
+  axom::ArrayView<const TetrahedronType> getCellsAsTets();
+  axom::ArrayView<const HexahedronType> getCellsAsHexes();
+  //!@brief Get volume of mesh cells.
+  axom::ArrayView<const double> getCellVolumes();
+  //!@brief Get characteristic lengths of mesh cells.
+  axom::ArrayView<const double> getCellLengths();
+  axom::ArrayView<const BoundingBox3DType> getCellBoundingBoxes();
+  axom::ArrayView<const IndexType, 2> getCellNodeConnectivity();
+  axom::ArrayView<const Point3DType> getVertexPoints();
+  const axom::StackArray<axom::ArrayView<const double>, 3>& getVertexCoords3D() const
+  {
+    return m_vertCoordsViews3D;
+  }
+  /*!
+   * @brief Precompute (and cache) mesh-dependent intermediate data
+   * that may be used in clipping.
+   *
+   * Mesh-dependent data are computed as needed and cached, but this
+   * computes them all at once.
+  */
+  void precomputeMeshData();
+  //@}
+
+  /*!
+   * @brief Check whether mesh meets requirements for shaping.
```

Comment:
Please document the requirements (at a high level)

Reply:
Good idea.  Done.